sopel/plugin.py

```diff
@@ -1809,19 +1815,48 @@ def handle_example_bugs(bot, trigger):
         :attr:`~sopel.config.core_section.CoreSection.auto_url_schemes`.
 
     """
-    def actual_decorator(function):
-        function._sopel_callable = True
-        if not hasattr(function, 'url_regex'):
-            function.url_regex = []
+    def decorator(
+        function: Callable | PluginGeneric,
+    ) -> PluginCallable:
+        # do we need to handle the match parameter?
+        # old style URL callback: callable(bot, trigger, match)
+        # new style: callable(bot, trigger)
+        # TODO: remove in Sopel 9
+        if isinstance(function, AbstractPluginObject):
+            url_handler = function.get_handler()
+        else:
+            url_handler = function
+
+        match_count = 3
+        if inspect.ismethod(url_handler):
+            # account for the 'self' parameter when the handler is a method
+            match_count = 4
+
+        argspec = inspect.getfullargspec(url_handler)
+
+        if len(argspec.args) >= match_count:
+            @functools.wraps(url_handler)
+            def wrapped_handler(bot, trigger):
+                return url_handler(bot, trigger, match=trigger)
+
+            if isinstance(function, AbstractPluginObject):
+                function._handler = wrapped_handler
```

Comment:
Pylance didn't like this when I opened VS Code to more easily look for where the `match` argument handling went.

In case it helps you further refine the WIP, here is the full table of problems I got (limited to open files, I believe, so this won't cover _everything_):

<img width="1740" height="500" alt="Image" src="https://github.com/user-attachments/assets/a6280b9c-705b-49d7-8765-4583d8ab12e6" />

Reply:
Yeah I tried to fix that, I think I'm not done yet. However, I've added a method to replace the handler, so it's now type safe.

Except that I went into a rabbit hole of type hinting things after the other since mypy started complaining about some stuffs here and there.

I'm not too worried, this is an improvement! More to come soon.